sysfs/class_infiniband.go

```diff
@@ -0,0 +1,372 @@
+	}
+	id = uint(value)
+	return id, name, nil
+}
+
+// Parse rate (example: "100 Gb/sec (4X EDR)") and return it as bytes/second
+func parseRate(s string) (uint64, error) {
+	var rate uint64
+	parts := strings.Split(s, "Gb/sec")
+	if len(parts) != 2 {
+		return rate, fmt.Errorf("failed to split '%s' by 'Gb/sec'", s)
+	}
+	value, err := strconv.ParseFloat(strings.TrimSpace(parts[0]), 32)
+	if err != nil {
+		return rate, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	rate = uint64(value * 125000000)
+	return rate, nil
+}
+
+// parseInfiniBandPort scans predefined files in /sys/class/infiniband/<device>/ports/<port>
+// directory and gets their contents.
+func (fs FS) parseInfiniBandPort(name string, port string) (*InfiniBandPort, error) {
+	portNumber, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert %s into uint", port)
+	}
+	infiniBandPort := InfiniBandPort{Name: name, Port: uint(portNumber)}
+
+	portPath := fs.sys.Path(infinibandClassPath, name, "ports", port)
+	content, err := ioutil.ReadFile(filepath.Join(portPath, "state"))
+	if err != nil {
+		return nil, err
+	}
+	id, name, err := parseState(string(content))
+	if err != nil {
+		return nil, fmt.Errorf("could not parse state file in %s: %s", portPath, err)
+	}
+	infiniBandPort.State = name
+	infiniBandPort.StateID = id
+
+	content, err = ioutil.ReadFile(filepath.Join(portPath, "phys_state"))
+	if err != nil {
+		return nil, err
+	}
+	id, name, err = parseState(string(content))
+	if err != nil {
+		return nil, fmt.Errorf("could not parse phys_state file in %s: %s", portPath, err)
+	}
+	infiniBandPort.PhysState = name
+	infiniBandPort.PhysStateID = id
+
+	content, err = ioutil.ReadFile(filepath.Join(portPath, "rate"))
+	if err != nil {
+		return nil, err
+	}
+	infiniBandPort.Rate, err = parseRate(string(content))
+	if err != nil {
+		return nil, fmt.Errorf("could not parse rate file in %s: %s", portPath, err)
+	}
+
+	counters, err := parseInfiniBandCounters(portPath)
+	if err != nil {
+		return nil, err
+	}
+	infiniBandPort.Counters = *counters
+
+	return &infiniBandPort, nil
+}
+
+func parseInfiniBandCounters(portPath string) (*InfiniBandCounters, error) {
+	var counters InfiniBandCounters
+
+	path := filepath.Join(portPath, "counters")
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file %q: %v", name, err)
+		}
+
+		// According to Mellanox, the metrics port_rcv_data, port_xmit_data,
+		// port_rcv_data_64, and port_xmit_data_64 "are divided by 4 unconditionally"
+		// as they represent the amount of data being transmitted and received per lane.
+		// Mellanox cards have 4 lanes per port, so all values must be multiplied by 4
```

Comment:
Does anyone other than Mellanox make infiniband gear?

Reply:
Mellanox (recently acquired by Nvidia) dominate the IB market by far, but previously Qlogic (tech acquired by Intel) and Voltaire (merged with Mellanox) produced IB hardware. Oracle produces their own IB chipsets.